### ISSUE TITLE
On `xvec_transform` and `plda`

### ISSUE THREAD

**Issue by alephpi** [BUTSpeechFIT/VBx#64]
Hi, me here again.
I'm referring to the operations:
`xvec_transform` as the following
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L125-L129

`plda` as the following
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L107-L113

I compared these logic with the paper Bayesian HMM clustering of x-vector sequences (VBx) in speaker diarization: Theory, implementation and analysis on standard tasks.

I think the `plda` part corresponds to (2)(3), but I cannot find any justification for `xvec_transform`, is it kind of preprocessing?
Also, when I look into
https://github.com/phonexiaresearch/VBx-training-recipe/blob/b699b5f1969e08faa7c0869d506b324613ef5c20/local/train_transform.py#L96-L132

I recognize it is again something to do with lda, right? **If I understand correctly, you have two LDA/PLDA process here, one for `xvec_transform` and the other for `plda` right?** Moreover, the output of `xvec_transform` is actually what you referred as untransformed x-vectors $\hat{\mathbf{X}}$ in the paper which is quite confusing at the first time.

I see clearly how you train the statistics for `xvec_transform` from `train_transform.py` but how do you train the statistics for `plda`, i.e. `plda_mu`,`plda_tr` and `plda_psi`? I have referred issue #32 but 
```
ivector-compute-plda ark:data/plda_train/spk2utt ark,cs:- exp/plda
```
**How do you do it? Where is the code?**

**Comment by fnlandini** [comment]
Hi,

I understand these steps are not very clear and they are rather more practical than related to the theory, which is why the paper is not very helpful.

These two transformations are somehow inherited from speaker recognition where the comparison of embeddings works in the same way. One can directly apply PLDA on the original space of the x-vectors and even reduce the dimensionality in the context of PLDA before calculating scores between the embeddings. However, it was seen in practice that reducing the dimensionality of the embeddings before using PLDA worked better. This is perhaps because of the Gaussianity assumption of PLDA so applying another transformation first could help with that.
One could use PCA for the first transformation, reduce the dimensionality and then use PLDA on this new space. Following standard practice in speaker recognition, the [first transformation](https://github.com/phonexiaresearch/VBx-training-recipe/blob/b699b5f1969e08faa7c0869d506b324613ef5c20/local/train_transform.py#L96-L132) is based on LDA.
As you pointed out, the parameters of LDA and PLDA are calculated in [this line](https://github.com/phonexiaresearch/VBx-training-recipe/blob/b699b5f1969e08faa7c0869d506b324613ef5c20/run.sh#L174), one after the other.
The calculation of the PLDA parameters relies on Kaldi through `ivector-compute-plda`. The key call is in [this line](https://kaldi-asr.org/doc/ivector-compute-plda_8cc_source.html#l00123) which iteratively estimates the parameters with the following calls:
```
void PldaEstimator::EstimateOneIter() {
	ResetPerIterStats();
	GetStatsFromIntraClass();
	GetStatsFromClassMeans();
	EstimateFromStats();
	KALDI_VLOG(2) << "Objective function is " << ComputeObjf();
}

void PldaEstimator::ResetPerIterStats() {
	within_var_stats_.Resize(Dim());
	within_var_count_ = 0.0;
	between_var_stats_.Resize(Dim());
	between_var_count_ = 0.0;
}

void PldaEstimator::GetStatsFromIntraClass() {
	within_var_stats_.AddSp(1.0, stats_.offset_scatter_);
	// Note: in the normal case, the expression below will be equal to the sum
	// over the classes, of (1-n), where n is the #examples for that class.  That
	// is the rank of the scatter matrix that "offset_scatter_" has for that
	// class. [if weights other than 1.0 are used, it will be different.]
	within_var_count_ += (stats_.example_weight_ - stats_.class_weight_);
}

void PldaEstimator::GetStatsFromClassMeans() {
	SpMatrix<double> between_var_inv(between_var_);
	between_var_inv.Invert();
	SpMatrix<double> within_var_inv(within_var_);
	within_var_inv.Invert();
	// mixed_var will equal (between_var^{-1} + n within_var^{-1})^{-1}.
	SpMatrix<double> mixed_var(Dim());
	int32 n = -1; // the current number of examples for the class.
 
	for (size_t i = 0; i < stats_.class_info_.size(); i++) {
		const ClassInfo &info = stats_.class_info_[i];
		double weight = info.weight;
		if (info.num_examples != n) {
			n = info.num_examples;
			mixed_var.CopyFromSp(between_var_inv);
			mixed_var.AddSp(n, within_var_inv);
			mixed_var.Invert();
		}
		Vector<double> m = *(info.mean); // the mean for this class.
		m.AddVec(-1.0 / stats_.class_weight_, stats_.sum_); // remove global mean
		Vector<double> temp(Dim()); // n within_var^{-1} m
		temp.AddSpVec(n, within_var_inv, m, 0.0);
		Vector<double> w(Dim()); // w, as defined in the comment.
		w.AddSpVec(1.0, mixed_var, temp, 0.0);
		Vector<double> m_w(m); // m - w
		m_w.AddVec(-1.0, w);
		between_var_stats_.AddSp(weight, mixed_var);
		between_var_stats_.AddVec2(weight, w);
		between_var_count_ += weight;
		within_var_stats_.AddSp(weight * n, mixed_var);
		within_var_stats_.AddVec2(weight * n, m_w);
		within_var_count_ += weight;
	}
}

void PldaEstimator::EstimateFromStats() {
	within_var_.CopyFromSp(within_var_stats_);
	within_var_.Scale(1.0 / within_var_count_);
	between_var_.CopyFromSp(between_var_stats_);
	between_var_.Scale(1.0 / between_var_count_);
 
	KALDI_LOG << "Trace of within-class variance is " << within_var_.Trace();
	KALDI_LOG << "Trace of between-class variance is " << between_var_.Trace();
}
```

I think this covers the second question but let me know if you meant something else.

PS: We will get back to you regarding the other issue soon.

**Comment by alephpi** [comment]
I want to ensure what is the input `ark:data/plda_train/spk2utt` of the `plda`? Is it the output of the `xvec_transform` together with its speaker label?

**Comment by fnlandini** [comment]
Hi,
[This line](https://kaldi-asr.org/doc/ivector-compute-plda_8cc_source.html#l00035) in `ivector-compute-plda` should be helpful. The two input parameters are the spk2utt and ivector. This follows the formats in Kaldi where spk2utt will have one line per speaker such that the line starts with the speaker identifier and then the utterance identifiers for all the utterances of that speaker follow. Something like the following, if I'm not mistaken
```
spk1 utt2 utt3 utt5
spk2 utt1 utt4
```
Then, the ivector file will have one line per utterance such that there is the utterance identifier followed by the ivector calculated over the speech of that utterance, something like
```
utt1 ivec1
utt2 ivec2
utt3 ivec3
utt4 ivec4
utt5 ivec5
```
where "ivec*" are actually the values or some structure containing the values of the embeddings.

The combination of these two things is the input, the first file saying which speaker corresponds to each utterance and the second file saying what ivector corresponds to each utterance.
One key feature in Kaldi is the possibility of piping the output of one call to the next one and that is what's happening with `ivector-compute-plda ark:data/plda_train/spk2utt ark,cs:- exp/plda` where the output of `train_transform.py` (the ivectors) is piped directly to the input of `ivector-compute-plda` by using `ark,cs:-`. Finally, `exp/plda` is the output directory where the PLDA is saved.

**Comment by alephpi** [comment]
Thank you!

**Comment by alephpi** [comment]
Hi, I wonder if there is a way to get the xvector of audio segments and the corresponding speaker labels you've used in training the plda model?

**Comment by fnlandini** [comment]
Hi, I am afraid I do not have them. You can follow this repository https://github.com/phonexiaresearch/VBx-training-recipe which has the scripts for generating the x-vectors and speaker labels (plus PLDA training).
I hope this helps

**Comment by alephpi** [comment]
Alright let me check it. BTW, do you still remember the performance that xvector extractor has achieved when you ended up training in terms of EER, cross-entropy,...?

**Comment by fnlandini** [comment]
Not really, maybe @Jamiroquai88 might remember
At the same time, those models are a few years old and newer x-vector extractors should be able to reach better performance. 
Still, as a reference, there is some information about EER and DER of some ResNet-based models in this recent [issue](https://github.com/BUTSpeechFIT/VBx/issues/65#issuecomment-1987959797)

**Comment by Jamiroquai88** [comment]
@alephpi sorry, I don't think I have anything useful
I have left the company where I worked on this and don't have any of those data. We trained this model specifically for diarization, we only did a quick test on how it performs on some standard conditions and it was pretty good as far as I remember (better than Resnet43, 18).

**Comment by alephpi** [comment]
> Not really, maybe @Jamiroquai88 might remember At the same time, those models are a few years old and newer x-vector extractors should be able to reach better performance. Still, as a reference, there is some information about EER and DER of some ResNet-based models in this recent [issue](https://github.com/BUTSpeechFIT/VBx/issues/65#issuecomment-1987959797)

I'm kind of confused by what you mean *x-vector* here. According to [this](https://arxiv.org/pdf/2005.07143.pdf) and [this](https://arxiv.org/pdf/1910.12592.pdf), the speaker embedding extracted by ResNet-based model is called *r-vector*. Do you mean by *x-vector* in a general sense that speaker embeddings extracted by TDNN structures, e.g. including the ECAPA-TDNN embeddings as well?

**Comment by fnlandini** [comment]
Yes, I mean x-vector in a loose sense as the embedding of any of these architectures where basically there are layers of some kind (like in TDNN, ResNet, ECAPA-TDNN, etc.) and then a pooling layer over the whole input as originally shown in [X-VECTORS: ROBUST DNN EMBEDDINGS FOR SPEAKER RECOGNITION](https://www.danielpovey.com/files/2018_icassp_xvectors.pdf)
This term has become common in the speaker verification field. Although some people used the term "r-vector", it did not become popular.
I'd say the main two variants are x-vector which has this pooling over the whole input and [d-vector](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41939.pdf) which in the context of diarization are usually extracted more regularly and have finer resolution. See [SPEAKER DIARIZATION WITH LSTM](https://arxiv.org/pdf/1710.10468.pdf)

**Comment by fnlandini** [comment]
Closing due to inactivity

**Comment by alephpi** [comment]
Hi I just wonder when doing plda pretraining, do you use the xvector extracted from augmented audio or only from the original one? I see in section 3.1.3 of the technical report, you extract one xvector every whole piece of recording, so you don't augement by clipping, but do you still augment by adding noise or reverberation?

Moreover, from my side I notice that even the original VoxCeleb1+2 (without VoxCeleb_CN) has over a million recordings, so do such huge amount of data train a plda make sense?

**Comment by fnlandini** [comment]
Hi @alephpi 
I did not take care of that part of the pipeline but according to the [training recipe](https://github.com/phonexiaresearch/VBx-training-recipe/blob/746190197958abdc8e8ffe5da6d9bb9d05797bc8/run.sh#L145) only one version of each piece is used (which can be the original or one of the augmented ones selected randomly). 
As for the number of recordings, perhaps it is possible to use a subset and still get reasonable performance. If you want to try that to speed up the process, I would recommend to drop utterances from speakers that have many of them so that you conserve as many speakers as possible.

**Comment by alephpi** [comment]
Hi, I don't understand https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L107-L113
when you read the plda statistics from file(trained and saved by kaldi), why you **reconstruct** the between-class and within-class matrix and **solve again** the generalized eigenvalue problem? Shouldn't the two matrices already be the solutions to the generalized eigenvalue problem as you commented here https://github.com/BUTSpeechFIT/VBx/blob/master/VBx/kaldi_utils.py#L25-L34
I just wonder why you doing this seemingly redundant step. However, when I check the new and old value of these two matrices, I find plda_psi(the eigenvalues) is the same, while plda_tr(the eigenvectors) is slightly different. Maybe it's something to do with numerical stability?

**Comment by fnlandini** [comment]
Hi @alephpi 
You are right that it is rather unnecessary now. I suspect that when I updated the code from a [previous version](https://github.com/BUTSpeechFIT/VBx/blob/94b98c6ccdef271e4087879571a35805091e44c2/diarization_PLDAadapt_AHCxvec_BHMMxvec.py#L83) I left it there. In that previous version, we showed how to perform an interpolation between an out-of-domain and an in-domain PLDAs for which these steps were necessary.

Sorry if that caused any confusion. As you pointed out, the differences between the two are very small indeed.

**Comment by alephpi** [comment]
Hi, I just want to make sure I understand the detail, 
here https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L113
and here https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L153
actually the **transpose** of `plda_tr` is the eigenmatrix, i.e whose column vectors are the eigenvectors of the generalized eigenvalue problem. So I guess that kaldi output `plda_tr` as they are row vectors, right?

**Comment by alephpi** [comment]
Sorry, but can you confirm the above message, it's really important for me to understand the behavior of the algorithm, since if I try to comment out https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/vbhmm.py#L109-L113
which are legacy codes as you said, it implies that the `plda_tr` here is actually the **transpose** of the eigenmatrix you've mentioned in formula (3) of the technical report.
